### PR TITLE
[MAINTENANCE] Standardize Store CRUD

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- Ensure that `add_store` always persists
- Add `delete_store`

### Definition of Done
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 1]
```diff
@@ -1193,15 +1193,40 @@ def add_store(self, store_name: str, store_config: dict) -> Optional[Store]:
...
+        Args:
+            store_name: The name of the Store to be deleted.
+
+        Raises:
+            StoreConfigurationError if the target Store is not found.
+        """
+        if store_name not in self.config.stores and store_name not in self._stores:
+            raise gx_exceptions.StoreConfigurationError(
+                f'Attempted to delete a store named: "{store_name}". It is not a configured store.'
+            )
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py:1222]: If it is in one, it should always be in the other

File: great_expectations/data_context/data_context/file_data_context.py [hunk 2]
```diff
@@ -135,15 +134,6 @@ def _init_variables(self) -> FileDataContextVariables:
         )
         return variables
 
-    def add_store(self, store_name: str, store_config: dict) -> Optional[Store]:
```
Comment (Kilo59) [inline great_expectations/data_context/data_context/file_data_context.py:138]: Any chance we can make the `add_store` on the `AbstractDC` never return `None`?

File: tests/data_context/test_data_context_state_management.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    EphemeralDataContext,
+)
+from great_expectations.data_context.types.base import (
+    DataContextConfig,
+    InMemoryStoreBackendDefaults,
+)
+from great_expectations.exceptions.exceptions import StoreConfigurationError
+
+
+class EphemeralDataContextSpy(EphemeralDataContext):
```
Comment (Kilo59) [inline tests/data_context/test_data_context_state_management.py:13]: Good use of a test double. 👍

File: tests/data_context/test_data_context_state_management.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        project_config: DataContextConfig,
+    ) -> None:
+        super().__init__(project_config)
+        self.save_count = 0
+
+    def _save_project_config(self):
+        """
+        No-op our persistence mechanism but increment an internal counter to ensure it was used.
+        """
+        self.save_count += 1
```
Comment (Kilo59) [inline tests/data_context/test_data_context_state_management.py:29]: If we ever move this into a `conftest.py` module I would suggest we log an `.info` level log message here.

Something very minimal but useful if someone is using this and doesn't realize it's just a `no-op` spy method.

**Review by Kilo59 — APPROVED** [review]
LGTM but I'd like to see another review since I'm not very familiar with `Stores`.

File: great_expectations/data_context/data_context/file_data_context.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/file_data_context.py:138]: yup modified the underlying helper that does this so we always return an obj

File: great_expectations/data_context/util.py [hunk 5]
```diff
@@ -109,9 +109,11 @@ def build_store_from_config(
     store_config: Optional[dict] = None,
     module_name: str = "great_expectations.data_context.store",
     runtime_environment: Optional[dict] = None,
-) -> Optional["Store"]:
+) -> "Store":
     if store_config is None or module_name is None:
-        return None
+        raise gx_exceptions.StoreConfigurationError(
+            "Cannot build a store without both a store_config and a module_name"
+        )
```
Comment (cdkini) [inline great_expectations/data_context/util.py]: Returning `None` here didn't make a lot of sense - we should fail and if the downstream caller wants to, they can use a `try/except`

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 6]
```diff
@@ -1189,19 +1188,44 @@ def _serialize_substitute_and_sanitize_datasource_config(
...
 
         Returns:
             store (Store)
         """
+        store = self._build_store_from_config(store_name, store_config)
 
-        self.config.stores[store_name] = store_config  # type: ignore[index]
-        return self._build_store_from_config(store_name, store_config)
+        # Both the config and the actual stores need to be kept in sync
+        self.config.stores[store_name] = store_config
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py:1204]: Generate config on request - this allows us to balance state between both objs

**Comment by Kilo59** [comment]
Github is being pretty uncool and not letting me comment on particular lines anymore.

Can you update the string `"Store"` return annotations to be non-strings with `from __future__ import annotations`?

**Pull request merged**